Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/distance.h

```diff
@@ -227,6 +378,313 @@ sample_triangles(const FaceRange& triangles,
+    {
+      if (nb_pts_l_u == 0)
+        nb_points = num_vertices(tm);
+      else
+        nb_points = static_cast<std::size_t>(
+          std::ceil( g.mesh_length()*nb_pts_a_u) );
+    }
+    this->out = CGAL::cpp11::copy_n(g, nb_points, this->out);
+  }
+  Randomizer get_randomizer()
+  {
+    return Randomizer(tm, pmap);
+  }
+  void internal_sample_triangles(double grid_spacing_, bool smpl_fcs, bool smpl_dgs)
+  {
+    this->out = sample_triangles<Geom_traits>(
+              faces(tm), tm, pmap, grid_spacing_, this->out,smpl_fcs, smpl_dgs, false);
+  }
+
+  std::size_t get_points_size()
+  {
+    return num_vertices(tm);
+  }
+};
+
+
+
+template<typename PointRange,
+         typename TriangleRange,
+         typename OutputIterator,
+         typename Geom_traits,
+         typename Creator,
+         typename NamedParameters>
+struct Triangle_structure_sampler_for_triangle_soup
+    : Triangle_structure_sampler_base<
+    OutputIterator,
+    Geom_traits,
+    NamedParameters,
+    typename TriangleRange::const_iterator,
+    Random_points_in_triangle_soup<PointRange, typename TriangleRange::value_type, Creator>,
+    Creator,
+    Triangle_structure_sampler_for_triangle_soup<PointRange,
+      TriangleRange,
+      OutputIterator,
+      Geom_traits,
+      Creator,
+      NamedParameters>
+    >
+{
+  typedef typename TriangleRange::value_type TriangleType;
+  typedef Triangle_structure_sampler_for_triangle_soup<PointRange,
+  TriangleRange,
+  OutputIterator,
+  Geom_traits,
+  Creator,
+  NamedParameters> This;
+
+  typedef Triangle_structure_sampler_base<
+  OutputIterator,
+  Geom_traits,
+  NamedParameters,
+  typename TriangleRange::const_iterator,
+  Random_points_in_triangle_soup<PointRange, TriangleType, Creator>,
+  Creator,
+  This> Base;
+
+
+
+  typedef Random_points_in_triangle_soup<PointRange, TriangleType, Creator> Randomizer;
+  typedef typename TriangleRange::const_iterator TriangleIterator;
+
+  double min_edge_length_;
+  const TriangleRange& triangles;
+  const PointRange& points;
+  Triangle_structure_sampler_for_triangle_soup(const PointRange& pts,
+                                               const TriangleRange& trs,
+                                               OutputIterator& out,
+                                               NamedParameters np)
+    :Base(out, np),
+      points(pts), triangles(trs)
+  {
+    using parameters::choose_parameter;
+    using parameters::get_parameter;
+
+    min_edge_length_ = (std::numeric_limits<double>::max)();
+  }
+
+  std::pair<TriangleIterator,TriangleIterator> get_range()
+  {
+    return std::make_pair(triangles.begin(), triangles.end());
+  }
+  void sample_points()
+  {
+    this->out = std::copy(
+          points.begin(),
+          points.end(),
+          this->out);
+  }
+  double get_minimum_edge_length()
+  {
+    if(min_edge_length_ != (std::numeric_limits<double>::max)())
+      return min_edge_length_;
+    for(auto tr : triangles)
```

Comment:
const&, you're copying everything there. Or just typedef the `value_type`, which is cleaner.